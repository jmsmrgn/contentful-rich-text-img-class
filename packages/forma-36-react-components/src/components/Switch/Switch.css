@import 'settings/colors';
@import 'settings/dimensions';
@import 'settings/transitions';

.Switch__wrapper {
  display: flex;
}

.Switch {
  box-sizing: content-box;
  width: calc(1rem * (34 / var(--font-base-default)));
  height: calc(1rem * (17 / var(--font-base-default)));
  background-color: var(--color-element-mid);
  border-radius: calc(1rem * (17 / var(--font-base-default)));
  cursor: pointer;
  outline: none;
  margin: 0 var(--spacing-xs) 0 0;
  border: 2px solid var(--color-element-mid);
  appearance: none;
  transition: background-color var(--transition-easing-default)
      var(--transition-duration-default),
    border-color var(--transition-easing-default)
      var(--transition-duration-default);
}

.Switch:before {
  content: '';
  display: block;
  width: 17px;
  height: 17px;
  background-color: var(--color-white);
  transition: transform var(--transition-easing-default)
    var(--transition-duration-default);
  border-radius: 50%;
  transform: translate3d(0, 0, 0);
}

.Switch--checked {
  background-color: var(--color-primary);
  border-color: var(--color-primary);
}

.Switch--disabled {
  background-color: var(--color-element-light);
  border-color: var(--color-element-light);
  cursor: not-allowed;
}

.Switch--checked.Switch--disabled {
  background-color: var(--color-element-mid);
  border-color: var(--color-element-mid);
}

.Switch--checked:before {
  transform: translate3d(100%, 0, 0);
}

.Switch__label {
  color: var(--color-text-dark);
}

.Switch__label--disabled {
  color: var(--color-text-light);
}
