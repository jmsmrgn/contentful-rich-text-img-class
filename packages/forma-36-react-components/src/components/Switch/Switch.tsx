import React, { FunctionComponent, KeyboardEvent } from 'react';
import FormLabel from '../FormLabel';
import classNames from 'classnames';
import styles from './Switch.css';

export interface SwitchProps {
  onToggle?: (isChecked: boolean) => void;
  isChecked?: boolean;
  isDisabled?: boolean;
  labelText: string;
  className?: string;
  id: string;
}

export const Switch: FunctionComponent<SwitchProps> = (props: SwitchProps) => {
  return (
    <div className={classNames(styles['Switch__wrapper'], props.className)}>
      <input
        type="checkbox"
        onClick={onToggle}
        onKeyUp={onKeyUp}
        defaultChecked={props.isChecked}
        disabled={props.isDisabled}
        className={classNames(styles['Switch'], {
          [styles['Switch--checked']]: props.isChecked,
          [styles['Switch--disabled']]: props.isDisabled,
        })}
        id={props.id}
      />
      <FormLabel
        className={classNames(styles['Switch__label'], {
          [styles['Switch__label--disabled']]: props.isDisabled,
        })}
        htmlFor={props.id}
      >
        {props.labelText}
      </FormLabel>
    </div>
  );

  function onToggle() {
    if (props.onToggle) {
      props.onToggle(!props.isChecked);
    }
  }

  function onKeyUp(e: KeyboardEvent<HTMLElement>) {
    if (props.onToggle && e.key === 'Enter') {
      props.onToggle(!props.isChecked);
    }
  }
};

export default Switch;
