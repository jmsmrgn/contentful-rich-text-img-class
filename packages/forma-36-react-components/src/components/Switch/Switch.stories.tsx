import React, { useState } from 'react';
import { storiesOf } from '@storybook/react';
import { text, boolean } from '@storybook/addon-knobs';

import Switch from './Switch';

function DefaultStory() {
  const [isActive, setActive] = useState(false);

  return (
    <Switch
      id="testSwitch"
      className={text('className', '')}
      isChecked={boolean('isChecked', isActive)}
      isDisabled={boolean('isDisabled', false)}
      labelText={text('labelText', 'My label text')}
      onToggle={setActive}
    />
  );
}

storiesOf('Components/Switch', module)
  .addParameters({
    propTypes: Switch['__docgenInfo'],
    component: Switch,
  })
  .add('default', () => <DefaultStory />);
