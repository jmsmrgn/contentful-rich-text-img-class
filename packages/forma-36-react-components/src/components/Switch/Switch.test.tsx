import React from 'react';
import { shallow } from 'enzyme';
import Switch from './Switch';

it('renders the component', () => {
  const output = shallow(<Switch id="testCheckbox" labelText="foobar" />);

  expect(output).toMatchSnapshot();
});

it('renders the component as checked', () => {
  const output = shallow(<Switch id="testCheckbox" labelText="foobar" isChecked />);

  expect(output).toMatchSnapshot();
});

it('renders the component as disabled', () => {
  const output = shallow(<Switch id="testCheckbox" labelText="foobar" isDisabled />);

  expect(output).toMatchSnapshot();
});

it('renders the component as disabled and checked', () => {
  const output = shallow(<Switch id="testCheckbox" labelText="foobar" isDisabled isChecked />);

  expect(output).toMatchSnapshot();
});

it('renders the component with an additional class name', () => {
  const output = shallow(
    <Switch id="testCheckbox" labelText="foobar" className="my-extra-class" />,
  );

  expect(output).toMatchSnapshot();
});
