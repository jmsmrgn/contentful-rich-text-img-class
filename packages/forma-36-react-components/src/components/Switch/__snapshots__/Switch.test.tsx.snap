// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the component 1`] = `
<div
  className="Switch__wrapper"
>
  <input
    className="Switch"
    id="testCheckbox"
    onClick={[Function]}
    onKeyUp={[Function]}
    type="checkbox"
  />
  <FormLabel
    className="Switch__label"
    htmlFor="testCheckbox"
    required={false}
    requiredText="required"
    testId="cf-ui-form-label"
  >
    foobar
  </FormLabel>
</div>
`;

exports[`renders the component as checked 1`] = `
<div
  className="Switch__wrapper"
>
  <input
    className="Switch Switch--checked"
    defaultChecked={true}
    id="testCheckbox"
    onClick={[Function]}
    onKeyUp={[Function]}
    type="checkbox"
  />
  <FormLabel
    className="Switch__label"
    htmlFor="testCheckbox"
    required={false}
    requiredText="required"
    testId="cf-ui-form-label"
  >
    foobar
  </FormLabel>
</div>
`;

exports[`renders the component as disabled 1`] = `
<div
  className="Switch__wrapper"
>
  <input
    className="Switch Switch--disabled"
    disabled={true}
    id="testCheckbox"
    onClick={[Function]}
    onKeyUp={[Function]}
    type="checkbox"
  />
  <FormLabel
    className="Switch__label Switch__label--disabled"
    htmlFor="testCheckbox"
    required={false}
    requiredText="required"
    testId="cf-ui-form-label"
  >
    foobar
  </FormLabel>
</div>
`;

exports[`renders the component as disabled and checked 1`] = `
<div
  className="Switch__wrapper"
>
  <input
    className="Switch Switch--checked Switch--disabled"
    defaultChecked={true}
    disabled={true}
    id="testCheckbox"
    onClick={[Function]}
    onKeyUp={[Function]}
    type="checkbox"
  />
  <FormLabel
    className="Switch__label Switch__label--disabled"
    htmlFor="testCheckbox"
    required={false}
    requiredText="required"
    testId="cf-ui-form-label"
  >
    foobar
  </FormLabel>
</div>
`;

exports[`renders the component with an additional class name 1`] = `
<div
  className="Switch__wrapper my-extra-class"
>
  <input
    className="Switch"
    id="testCheckbox"
    onClick={[Function]}
    onKeyUp={[Function]}
    type="checkbox"
  />
  <FormLabel
    className="Switch__label"
    htmlFor="testCheckbox"
    required={false}
    requiredText="required"
    testId="cf-ui-form-label"
  >
    foobar
  </FormLabel>
</div>
`;
